bt_navigator:
  ros__parameters:
    use_sim_time: False
    global_frame: map
    robot_base_frame: base_footprint
    odom_topic: /final_pose_nav
    bt_loop_duration: 10
    default_server_timeout: 20
    wait_for_service_timeout: 1000
    # 'default_nav_through_poses_bt_xml' and 'default_nav_to_pose_bt_xml' are use defaults:
    # nav2_bt_navigator/navigate_to_pose_w_replanning_and_recovery.xml
    # nav2_bt_navigator/navigate_through_poses_w_replanning_and_recovery.xml
    # They can be set here or via a RewrittenYaml remap from a parent launch file to Nav2.
    default_nav_through_poses_bt_xml: "$(find-pkg-share custom_nav_thru_poses)/behavior_trees/navigate_through_poses_w_selector.xml"
    default_nav_to_pose_bt_xml: "$(find-pkg-share custom_nav_to_pose)/behavior_trees/navigate_to_pose_w_selector.xml"
    plugin_lib_names:
      - nav2_stop_controller_bt_node
      - nav2_compute_path_to_pose_action_bt_node
      - nav2_compute_path_through_poses_action_bt_node
      - nav2_smooth_path_action_bt_node
      - nav2_follow_path_action_bt_node
      - nav2_spin_action_bt_node
      - nav2_wait_action_bt_node
      - nav2_escape_action_bt_node
      - nav2_shrink_action_bt_node
      - nav2_assisted_teleop_action_bt_node
      - nav2_back_up_action_bt_node
      - nav2_drive_on_heading_bt_node
      - nav2_clear_costmap_service_bt_node
      - nav2_is_stuck_condition_bt_node
      - nav2_goal_reached_condition_bt_node
      - nav2_goal_updated_condition_bt_node
      - nav2_globally_updated_goal_condition_bt_node
      - nav2_is_path_valid_condition_bt_node
      - nav2_initial_pose_received_condition_bt_node
      - nav2_reinitialize_global_localization_service_bt_node
      - nav2_rate_controller_bt_node
      - nav2_distance_controller_bt_node
      - nav2_speed_controller_bt_node
      - nav2_truncate_path_action_bt_node
      - nav2_truncate_path_local_action_bt_node
      - nav2_goal_updater_node_bt_node
      - nav2_recovery_node_bt_node
      - nav2_pipeline_sequence_bt_node
      - nav2_round_robin_node_bt_node
      - nav2_transform_available_condition_bt_node
      - nav2_time_expired_condition_bt_node
      - nav2_path_expiring_timer_condition
      - nav2_distance_traveled_condition_bt_node
      - nav2_single_trigger_bt_node
      - nav2_goal_updated_controller_bt_node
      - nav2_is_battery_low_condition_bt_node
      - nav2_navigate_through_poses_action_bt_node
      - nav2_navigate_to_pose_action_bt_node
      - nav2_remove_passed_goals_action_bt_node
      - nav2_planner_selector_bt_node
      - nav2_controller_selector_bt_node
      - nav2_goal_checker_selector_bt_node
      - nav2_controller_cancel_bt_node
      - nav2_path_longer_on_approach_bt_node
      - nav2_wait_cancel_bt_node
      - nav2_spin_cancel_bt_node
      - nav2_back_up_cancel_bt_node
      - nav2_assisted_teleop_cancel_bt_node
      - nav2_drive_on_heading_cancel_bt_node
      - nav2_is_battery_charging_condition_bt_node

ControllerSelector:
  ros__parameters:
    Fast_controller: "Fast"
    Slow_controller: "Slow"

bt_navigator_navigate_through_poses_rclcpp_node:
  ros__parameters:
    use_sim_time: False

bt_navigator_navigate_to_pose_rclcpp_node:
  ros__parameters:
    use_sim_time: False

controller_server:
  ros__parameters:
    use_sim_time: False
    holonomic: true
    use_omnidirectional_robot: true
    controller_frequency: 20.0
    min_x_velocity_threshold: 0.001
    min_y_velocity_threshold: 0.001
    min_theta_velocity_threshold: 0.001
    failure_tolerance: 0.0
    progress_checker_plugin: "progress_checker"
    goal_checker_plugins: ["Precise", "Unerring", "Loose"]
    controller_plugins: ["Fast", "Slow", "LinearBoost", "AngularBoost"]

    # Progress checker parameters
    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.03
      movement_time_allowance: 3.0

    # Goal checker parameters
    Precise:
      stateful: True
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.01
      yaw_goal_tolerance: 0.05
      xy_debounce_threshold: 1
      yaw_debounce_threshold: 1

    Unerring:
      stateful: True
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.01
      yaw_goal_tolerance: 0.05
      xy_debounce_threshold: 5
      yaw_debounce_threshold: 5

    Loose:
      stateful: True
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.03
      yaw_goal_tolerance: 0.07
      xy_debounce_threshold: 1
      yaw_debounce_threshold: 1

    # Controller parameters
    Fast:
      plugin: "custom_controller::CustomController"
      max_linear_vel: 0.5
      max_angular_vel: 2.0
      linear_kp: 1.5
      angular_kp: 3.0
      transform_tolerance: 0.1
      look_ahead_distance: 0.2
      speed_decade: 0.125
      costmap_tolerance: 95
      keep_planning: false
      
    Slow:
      plugin: "custom_controller::CustomController"
      max_linear_vel: 0.5
      max_angular_vel: 2.0
      linear_kp: 1.5
      angular_kp: 3.0
      transform_tolerance: 0.1
      look_ahead_distance: 0.2
      speed_decade: 0.125
      costmap_tolerance: 95
      keep_planning: false

    LinearBoost:
      plugin: "custom_controller::CustomController"
      max_linear_vel: 0.5
      max_angular_vel: 2.0
      linear_kp: 1.5
      angular_kp: 3.0
      transform_tolerance: 0.1
      look_ahead_distance: 0.2
      speed_decade: 0.125
      costmap_tolerance: 95
      keep_planning: false

    AngularBoost:
      plugin: "custom_controller::CustomController"
      max_linear_vel: 0.5
      max_angular_vel: 2.0
      linear_kp: 1.5
      angular_kp: 3.0
      transform_tolerance: 0.1
      look_ahead_distance: 0.2
      speed_decade: 0.125
      costmap_tolerance: 95
      keep_planning: false

local_costmap:
  local_costmap:
    ros__parameters:
      update_frequency: 20.0
      publish_frequency: 20.0
      global_frame: map
      robot_base_frame: base_footprint
      rolling_window: False
      width: 3
      height: 2
      resolution: 0.01
      robot_radius: 0.22
      # plugins: ["static_layer", "inflation_layer"]
      plugins: ["static_layer"]
      static_layer:
        plugin: "nav2_costmap_2d::StaticLayer"
        map_subscribe_transient_local: True
      always_send_full_costmap: True

global_costmap:
  global_costmap:
    ros__parameters:
      update_frequency: 20.0
      publish_frequency: 20.0
      global_frame: map
      robot_base_frame: base_footprint
      use_sim_time: False
      robot_radius: 0.22
      resolution: 0.01
      track_unknown_space: false
      trinary_costmap: false
      lethal_cost_threshold: 10
      unknown_cost_value: 255
      plugins: ["static_layer", "inflation_layer", "rival_layer", "object_layer", "keepout_layer"]
      static_layer:
        plugin: "nav2_costmap_2d::StaticLayer"
        map_subscribe_transient_local: True
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 5.0
        inflation_radius: 0.2
      rival_layer:
        plugin: "custom_path_costmap_plugin::RivalLayer"
        enabled: True
        model_size: 22
        x_cov_threshold: 0.02
        y_cov_threshold: 0.02
        R_sq_threshold: 0.85
        reset_timeout_threshold: 70
        robot_inscribed_radius: 0.3
        external_rival_data_path: "/home/share/data/rival_param.yaml"
        rival_inscribed_radius: 0.22
        halted_inflation_radius: 0.1
        wandering_inflation_radius: 0.12
        moving_inflation_radius: 0.1
        unknown_inflation_radius: 0.12
        halted_cost_scaling_factor: 10.0
        wandering_cost_scaling_factor: 10.0
        moving_cost_scaling_factor: 0.9
        unknown_cost_scaling_factor: 10.0
        max_extend_length: 0.6
        cov_range_max: 0.05385164807134504  # sqrt(0.0029)
        cov_range_min: 0.01414213562373095  # sqrt(0.0002)
        vel_range_max: 1.0
        vel_range_min: 0.0
        inscribed_radius_rate: 0.99
        inflation_radius_rate: 1.005
        debug_mode: 0                       # 0: off, 1: Print rival state change only, 2: Print rival state change and statistics, 3: Print everything continuously
        use_statistic_method: False
        offset_vel_factor_weight_statistic: 0.42
        expand_vel_factor_weight_statistic: 0.42
        offset_vel_factor_weight_localization: 0.42
        expand_vel_factor_weight_localization: 0.42
        safe_distance: 0.5
      object_layer:
        plugin: "Object_costmap_plugin::ObjectLayer"
        enabled: True
        column_inscribed_radius: 0.075
        board_inscribed_radius: 0.01
        column_inflation_radius: 0.22
        board_inflation_radius: 0.22
        cost_scaling_factor: 3.0
      keepout_layer:
        plugin: "keepout_costmap_plugin::KeepoutLayer"
        enabled: True
        # keepout_zone_array: [X, Y, lengthX, lengthY]
        keepout_zone_array: [0.375, 1.775, 0.45, 0.45,  # A
                             0.225, 0.875, 0.45, 0.45,  # B
                             0.225, 0.075, 0.45, 0.15,  # C
                             0.775, 0.075, 0.45, 0.15,  # D
                             1.225, 0.225, 0.45, 0.45,  # E
                             #----------------------------#
                             1.775, 0.225, 0.45, 0.45,  # F
                             2.225, 0.075, 0.45, 0.15,  # G
                             2.775, 0.075, 0.45, 0.15,  # H
                             2.775, 0.875, 0.45, 0.45,  # I
                             2.625, 1.775, 0.45, 0.45,] # J 
        inflation_length: 0.2
        cost_scaling_factor: 5.0
      always_send_full_costmap: True

map_server:
  ros__parameters:
    use_sim_time: False
    # Overridden in launch by the "map" launch configuration or provided default value.
    # To use in yaml, remove the default "map" value in the tb3_simulation_launch.py file & provide full path to map below.
    yaml_filename: "basic_map.yaml"

planner_server:
  ros__parameters:
    expected_planner_frequency: 20.0
    use_sim_time: False
    use_omnidirectional_robot: true
    planner_plugins: ["GridBased"]
    GridBased:
      plugin: "nav2_navfn_planner/NavfnPlanner"
      tolerance: 0.02
      use_astar: True
      allow_unknown: false

smoother_server:
  ros__parameters:
    use_sim_time: False
    smoother_plugins: ["simple_smoother"]
    simple_smoother:
      plugin: "nav2_smoother::SimpleSmoother"
      tolerance: 1.0e-10
      max_its: 1000
      do_refinement: True

behavior_server:
  ros__parameters:
    costmap_topic: local_costmap/costmap_raw
    footprint_topic: local_costmap/published_footprint
    cycle_frequency: 10.0
    behavior_plugins: ["wait", "escape", "shrink"]
    escape:
      plugin: "nav2_behaviors/Escape"
    wait:
      plugin: "nav2_behaviors/Wait"
    shrink:
      plugin: "nav2_behaviors/Shrink"
    global_frame: map
    robot_base_frame: base_footprint
    transform_tolerance: 0.1
    use_sim_time: False
    simulate_ahead_time: 2.0
    max_rotational_vel: 1.0
    min_rotational_vel: 0.4
    rotational_acc_lim: 3.2

robot_state_publisher:
  ros__parameters:
    use_sim_time: False

waypoint_follower:
  ros__parameters:
    use_sim_time: False
    loop_rate: 20
    stop_on_failure: false
    waypoint_task_executor_plugin: "wait_at_waypoint"
    wait_at_waypoint:
      plugin: "nav2_waypoint_follower::WaitAtWaypoint"
      enabled: True
      waypoint_pause_duration: 200

velocity_smoother:
  ros__parameters:
    use_sim_time: False
    smoothing_frequency: 20.0
    scale_velocities: False
    feedback: "OPEN_LOOP"
    max_velocity: [1.1, 1.1, 12.0]
    min_velocity: [-1.1, -1.1, -12.0]
    max_accel: [3.3, 3.3, 120.0]
    max_decel: [-4.4, -4.4, -120.0]  
    odom_topic: "/final_pose_nav"
    odom_duration: 0.1
    deadband_velocity: [0.0, 0.0, 0.0]
    velocity_timeout: 0.1 # must be less than velocity_smoother_wait in docking_server

docking_server:
  ros__parameters:
    controller_frequency: 20.0
    initial_perception_timeout: 5.0
    wait_charge_timeout: 5.0
    dock_approach_timeout: 3.0
    max_retries: 0
    base_frame: "base_footprint"
    fixed_frame: "map"
    dock_backwards: false
    dock_prestaging_tolerance: 0.01
    velocity_smoother_wait: 0.12

    # Types of docks
    dock_plugins: ["dock"]
    dock:
      plugin: 'opennav_docking::SimpleChargingDock'
      docking_threshold: 0.005
      use_debounce: true
      xy_debounce_threshold: 2
      yaw_debounce_threshold: 2
      use_dynamic_offset: true
      staging_x_offset: 0.35
      staging_y_offset: 0.35
      staging_yaw_offset: 1.57
      use_battery_status: false
      use_stall_detection: false
      base_frame: "base_footprint"

    docks: ['home_dock']
    home_dock:
      type: 'dock'
      frame: map
      pose: [0.0, 0.0, 0.0]
    
    # dock_database: /my/path/to/dock_database.yaml

    controllers: ["Ordinary", "Gentle", "Rush"]

    Ordinary:
      max_linear_vel: 0.3
      min_linear_vel: 0.0
      linear_kp_accel_vel: 0.5
      linear_ki_accel_vel: 0.1
      linear_kp_decel_dis: 3.0
      angular_kp: 4.0
      deceleration_distance: 0.05
      reserved_distance: 0.0
      stop_degree: 120.0
      rival_radius: 0.44

    Gentle:
      max_linear_vel: 0.3
      min_linear_vel: 0.0
      linear_kp_accel_vel: 0.5
      linear_ki_accel_vel: 0.1
      linear_kp_decel_dis: 3.0
      angular_kp: 4.0
      deceleration_distance: 0.05
      reserved_distance: 0.0
      stop_degree: 120.0
      rival_radius: 0.44

    Rush:
      max_linear_vel: 0.3
      min_linear_vel: 0.0
      linear_kp_accel_vel: 0.5
      linear_ki_accel_vel: 0.1
      linear_kp_decel_dis: 3.0
      angular_kp: 4.0
      deceleration_distance: 0.05
      reserved_distance: 0.0
      stop_degree: 120.0
      rival_radius: 0.44

final_pose_bridge:
  ros__parameters:
    # twist_topic: "/driving_duaiduaiduai"
    twist_topic: "/odoo_googoogoo"